UW433: Forecast docs (#389)
ufs-community/uwtools: docs/sections/user_guide/uw_yaml/forecast_yaml.rst

.. _forecast_yaml:

Forecast YAML
=============

The structured YAML to run a forecast is described below. It is enforced via JSON Schema.

In this section, there are entries that define where the forecast will run (``run_directory:``), what will run (``executable:``), the data files that should be staged (``cycle-dependent:`` and ``static:``), and blocks that correspond to the configuration files required by the forecast model (``fd_ufs:``, ``field_table:``, ``namelist:``, etc.). Each of the configuration file blocks will allow the user to set a template file as input, or to define a configuration file in its native key/value pair format with an option to update the values (``update_values:``) contained in the original input file (``base_file:``).

> The configuration files required by the ufs-weather-model are documented :weather-model-io:`here<model-configuration-files>`.

Reviewer: ```suggestion
The configuration files required by the UFS Weather Model are documented :weather-model-io:`here<model-configuration-files>`.
```
Not sure if I'm right in suggesting this change -- the linked-to page has it as "UFS Weather Model", FWIW.

Author: Yep. That's good. I was stuck in "repo name" mode here.

The ``forecast:`` section
-------------------------

> The ``forecast:`` section describes the specifics of the FV3 atmosphere forecast component.

Reviewer: ```suggestion
This section describes the specifics of the FV3 atmosphere forecast component.
```
Slightly DRYer, since `forecast:` is right there in the section heading.

      INPUT/gfs_ctrl.nc: /path/to/gfs_ctrl.nc
    diag_table:
      template_file: /path/to/diag_table/template/file
    domain: regional
    executable: fv3.exe
    fd_ufs:
      base_file: /path/to/base/fd_ufs.yaml
    field_table:
      base_file: /path/to/field_table.yaml
      update_values:
        liq_wat:
          longname: cloud water mixing ratio
          units: kg/kg
          profile_type:
            name: fixed
            surface_value: 2.0
    length: 12
    model_configure:
      base_file: /path/to/base/model_configure
      update_values:
        write_dopost: .false.
    namelist:
      base_file: /path/to/base/input.nml
      update_values:
        fv_core_nml:
          k_split: 2
          n_split: 6
    nems.configure:
      template_file: /path/to/template/nems.configure
    run_dir: /path/to/forecast/run/directory/{yyyymmddhh}
    static:
      fv3.exe: /path/to/executable/ufs_model
      INPUT/grid_spec.nc: /path/to/grid_spec.nc
      ...
      data_table: /path/to/data_table
      eta_micro_lookup.data: /path/to/noahmptable.dat
      noahmptable.tbl: /path/to/noahmptable.tbl

.. _updating_values:

Updating Values
^^^^^^^^^^^^^^^

> Many of the sections describe the configuration files needed by the FV3 Forecast model, i.e. namelist, fd_ufs, model_configure. The ``base_file:`` entry is required so that the file will be staged. Modifications to the base file can be made at run time by providing an ``update_values:`` block.

Reviewer: ```suggestion
Many of the sections describe configuration files needed by FV3, e.g. ``namelist:``, ``fd_ufs:``, ``model_configure:``. The ``base_file:`` section is required to initially stage the file; it can then be modified via an ``update_values:`` block.
```
There's a bit of variance throughout this page in terminology -- `block`, `entry`, `section`. It would be helpful to harmonize them so that readers can count on a clear, unambiguous vocabulary.

Author: Just to make sure I am clear on those, here's what I've been assuming: 

`block` corresponds to a nested dict in Python terminology, `entry` is a single key/value pair, and `section` is a top-level `block`.

Reviewer: I think the highest-level block can just be called "block" -- I don't think it's ambiguous to say "the `forecast` block" or "the `workflow` block". Even if those keys happen to be repeated somewhere else in a huge UW YAML key, I think it will be clear from context which one we're talking about.

> To ensure the correct values are updated, all keys that act as section headers above the entry to be updated need to be provided in the order in which they appear in the base file. Multiple entries within a block may be updated and they need not follow the same order as those in the base file. For example, the base file named ``people.yaml`` may contain:

Author: Ahhh. Yeah, we were trying to figure out the correct wording here and tried to exemplify what we meant. Essentially, you need to mirror the structure from the base file _exactly_ to get the correct value to update.

> To ensure the correct values are updated, all keys that act as section headers above the entry to be updated need to be provided in the order in which they appear in the base file. Multiple entries within a block may be updated and they need not follow the same order as those in the base file. For example, the base file named ``people.yaml`` may contain:

Reviewer: I believe you, but that seems wrong. Why would the order matter?

> To ensure the correct values are updated, all keys that act as section headers above the entry to be updated need to be provided in the order in which they appear in the base file. Multiple entries within a block may be updated and they need not follow the same order as those in the base file. For example, the base file named ``people.yaml`` may contain:

Reviewer: Thanks, I understand now. I think "hierarchy" and/or "structure" would say that better than "order".

.. code-block:: yaml

   person:
     age: 19
     address:
       city: Boston
       number: 12
       state: MA
       street: Acorn St
     name: Jane

Then the entries under the YAML section would edit this base file with the entries:

.. code-block:: yaml

   base_file: people.yaml
   update_values:
     person:
       address:
         street: Main St
         number: 99

Rendering Template Files
^^^^^^^^^^^^^^^^^^^^^^^^

> Requires a path to a template file in the ``template_file:`` entry. There is no option to add values in the YAML. Instead, the driver is programmed to enter necessary values for the template.

Author: Here, I was looking for a way to say that the driver owns the values needed for the templates. The user doesn't provide them directly as they might in the `update_values:` entries for a YAML or namelist file.

> Requires a path to a template file in the ``template_file:`` entry. There is no option to add values in the YAML. Instead, the driver is programmed to enter necessary values for the template.

Reviewer: That makes me wonder: If there's no extra information to be provided, could we just have

```
diag_table: /path/to/diag_table/template/file
```

instead of

```
diag_table:
  template_file: /path/to/diag_table/template/file
```

? Is the extra level providing anything except potential confusion such that we have to document for the user (I'm paraphrasing what I'm now guessing is the purpose of this bit of doc), "Don't worry that you're not seeing extra keys and values here!"

> Requires a path to a template file in the ``template_file:`` entry. There is no option to add values in the YAML. Instead, the driver is programmed to enter necessary values for the template.

Author: It was originally meant to provide a visual cue that it was treated as a template. And it might be helpful to one day build in an option to override values provided to the template. (I'm screaming YAGNI at myself right now).

> Requires a path to a template file in the ``template_file:`` entry. There is no option to add values in the YAML. Instead, the driver is programmed to enter necessary values for the template.

Reviewer: I'm glad I don't have to :)

UW YAML Keys
^^^^^^^^^^^^

``cycle-dependent:``
""""""""""""""""""""

This block contains a set of files to stage in the run directory: file names as they appear in the run directory are keys and their source paths are the values. Source paths can be provided as a single string path, or a list of paths to be staged in a common directory under their original names.

  .. warning:: The beta version does not support adding cycle information to the content of the files, and this information must be hard-coded in the YAML file.

``diag_table:``
"""""""""""""""

In ``uwtools``, the ``diag_table`` is treated as a template so that the date and time information in the header may be filled in appropriately. The ``template_file:`` is the path to the input Jinja2 template. Date information is provided in the command line or API interfaces.

The diag_table is described :weather-model-io:`here<diag-table-file>`.

``domain:``
"""""""""""

> A switch to differentiate between a global or regional configuration. Accepted values are ``global`` and ``regional``.

Author: I added `domain` to the JSON schema as a string with an enum. Is there another step I missed?

> A switch to differentiate between a global or regional configuration. Accepted values are ``global`` and ``regional``.

Reviewer: Comparing the `FV3Forecast.jsonschema` in this PR to the YAML block [here](https://uwtools--389.org.readthedocs.build/en/389/sections/user_guide/uw_yaml/forecast_yaml.html#the-forecast-section), keys I do not see in the schema are: `diag_table`, `template_file`, `fd_ufs`, `nems.configure` (which I'd like to see as `nems_configure`).

I'd think that any key that we know in advance should be defined in the schema. Obviously we can't do keys that correspond to, say, namelist variables that we can't know in advance (because a model developer could add new ones, and will not be in a position to edit JSON Schema to get `uw` to accept those new variables).

I also think that we should define the structure of `field_table.update_values` further. If I understood the docs I read, anyway, we know that the structure has to contain `longname`, `units`, etc. keys -- even though we don't know the key name, which is fine because we can define _that_ with a regex. But the sub-structure is fixed, right? It will be very helpful for our users if we can get them as close as possible to correctness via schema checking.

I also don't see `additionalProperties: false` set anywhere here, so users can confuse themselves with misspellings, adding things at the wrong level, etc. I think we should be setting `additionalProperties: false` for every `object` where the set of acceptable keys is fixed, and we should define those keys.

So, there's that, in terms of completeness. I think this schema has been languishing for a long time while we improved `rocoto.jsonschema`. It's probably time to renovate this one using some best practices we've learned.

Apologies if I'm missing something about why this schema can't be locked down more.

One last thing: I've been adding [schema tests](https://uwtools--389.org.readthedocs.build/en/389/sections/user_guide/uw_yaml/forecast_yaml.html#the-forecast-section) for the Rocoto schema as a way to test very granular schema correctness without having to put together big YAML blocks. We may want to do the same for the FV3 schema.

Possibly all of this should be a separate ticket, or multiple tickets. Assuming I'm not wrong about it being desired/needed...

> A switch to differentiate between a global or regional configuration. Accepted values are ``global`` and ``regional``.

Author: It would be lovely to get that in a 2nd go, if you're okay with me capturing that in the Jira backlog.

> A switch to differentiate between a global or regional configuration. Accepted values are ``global`` and ``regional``.

Reviewer: Good plan.

``static:``
"""""""""""

This block contains a set of files to stage in the run directory: file names as they appear in the run directory are keys and their source paths are the values. Source paths can be provided as a single string path, or a list of paths to be staged in a common directory under their original names.

The ``platform:`` section
-------------------------

The ``platform:`` section describes necessary facts about the computational platform.

.. code-block:: yaml

  platform:
    mpicmd: srun # required
    scheduler: slurm

``mpicmd:``
^^^^^^^^^^^
The MPI command used to run the model executable. Typical options are ``srun``, ``mpirun``, ``mpiexec``, etc. System administrators should be able to advise the appropriate choice, if needed.

``scheduler:``
^^^^^^^^^^^^^^
The name of the batch system. Supported options are ``lfs``, ``pbs``, and ``slurm``.

The ``preprocessing:`` section
------------------------------

.. code-block:: yaml

  preprocessing:
    lateral_boundary_conditions:
      interval_hours: 3 # optional, default
      offset: 0 # optional, default
      output_file_path: # required

> ``lateral_boundary_conditions:``

Reviewer: Oh, sorry, I see the difference in the formatting marks now.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The optional section describes how the lateral boundary conditions have been prepared for a limited-area configuration of the model forecast. It is required for a limited-area forecast. The following entries in its subtree are used for the forecast:

``interval_hours:``
"""""""""""""""""""
The integer number of hours setting how frequently the lateral boundary conditions will be used in the model forecast.

``offset:``
"""""""""""

> The integer number of hours indicates how many hours earlier the external model used for boundary conditions started compared to the desired forecast cycle.

Reviewer: `indicates` -> `indicating`, at least.

But also, above, we say "setting", and here "indicating". Just my opinion, but: Unlike in normal prose, where a variety of terms for the same thing are very welcome because they're colorful and break the monotony, technical docs benefit from repetition of precise, familiar terms that readers can absorb and then never wonder about. They're looking for the juicy facts, which should stand out from the least amount of surrounding language needed to convey those facts. Maybe that's just my brain and me as a reader! But as a user I appreciate technical documentation that uses consistent, well-defined terms and as little extra language as possible. (Easy to say -- often hard to avoid "extra language".)

      update_values:
        liq_wat:
          longname: cloud water mixing ratio
          units: kg/kg
          profile_type:
            name: fixed
            surface_value: 2.0
    length: 12
    model_configure:
      base_file: /path/to/base/model_configure
      update_values:
        write_dopost: .false.
    namelist:
      base_file: /path/to/base/input.nml
      update_values:
        fv_core_nml:
          k_split: 2
          n_split: 6
    nems.configure:
      template_file: /path/to/template/nems.configure
    run_dir: /path/to/forecast/run/directory/{yyyymmddhh}
    static:
      fv3.exe: /path/to/executable/ufs_model
      INPUT/grid_spec.nc: /path/to/grid_spec.nc
      ...
      data_table: /path/to/data_table
      eta_micro_lookup.data: /path/to/noahmptable.dat
      noahmptable.tbl: /path/to/noahmptable.tbl

.. _updating_values:

Updating Values
^^^^^^^^^^^^^^^

Many of the blocks describe configuration files needed by the UFS Weather Model, i.e. ``namelist:``, ``fd_ufs:``, ``model_configure:``. The ``base_file:`` entry in a given block is required to initially stage the file; it can then be modified via an ``update_values:`` block.

> To ensure the correct values are updated, the hierarchy of entries in the base file must be mirrord under the ``update_values:`` block. Multiple entries within a block may be updated and they need not follow the same order as those in the base file. For example, the base file named ``people.yaml`` may contain:

Reviewer 2: mirrord -> mirrored

The diag_table is described :weather-model-io:`here<diag-table-file>`.

``domain:``
"""""""""""

A switch to differentiate between a global or regional configuration. Accepted values are ``global`` and ``regional``.

``executable:``
"""""""""""""""

The path to the compiled executable.

``fd_ufs:``
""""""""""""

This block requires a ``base_file:`` entry that contains the path to the YAML file. An optional ``update_values:`` block may be provided to update any values contained in the base file. Please see the :ref:`updating_values` section for providing information in these entries.

The ``fd_ufs.yaml`` file is a structured YAML used by the FV3 weather model. The tested version can be found in the :ufs-weather-model:`ufs-weather-model repository<blob/develop/tests/parm/fd_ufs.yaml>`. The naming convention for the dictionary entries are documented :cmeps:`here<>`.

``field_table:``
""""""""""""""""

The block requires a ``base_file:`` entry that contains the path to the YAML file. An optional ``update_values:`` block may be provided to update any values contained in the base file. Please see the :ref:`updating_values` section for providing information in these entries.

> If a pre-defined field table (i.e., not a configurable YAML) is to be used, include it in the ``static:`` block.

Reviewer 2: pre-defined -> predefined
